votor/src/voting_utils.rs

```diff
@@ -313,3 +311,293 @@ pub fn generate_vote_message(
+            message,
+            slot,
+            saved_vote_history,
+        } = &result
+        {
+            assert_eq!(slot, &vote_slot);
+            assert_eq!(**message, expected_message);
+            assert_eq!(
+                saved_vote_history,
+                &SavedVoteHistoryVersions::from(
+                    SavedVoteHistory::new(
+                        &voting_context.vote_history,
+                        &voting_context.identity_keypair
+                    )
+                    .unwrap()
+                )
+            );
+        } else {
+            panic!("Expected BLSOp::PushVote, got {result:?}");
+        }
+
+        // Check that own vote sender receives the vote
+        let received_message = own_vote_receiver.recv().unwrap();
+        assert_eq!(received_message, expected_message);
+    }
+
+    #[test]
+    fn test_wrong_vote_account() {
+        solana_logger::setup();
+        let (own_vote_sender, _own_vote_receiver) = crossbeam_channel::unbounded();
+        // Create 10 node validatorvotekeypairs vec
+        let validator_keypairs = (0..10)
+            .map(|_| ValidatorVoteKeypairs::new(Keypair::new(), Keypair::new(), Keypair::new()))
+            .collect::<Vec<_>>();
+        let my_index = 0;
+        let mut voting_context =
+            setup_voting_context_and_bank_forks(own_vote_sender, &validator_keypairs, my_index);
+
+        // If we haven't reached wait_to_vote_slot yet, return Ok(None)
+        voting_context.wait_to_vote_slot = Some(4);
+        let vote = Vote::new_finalization_vote(2);
+        assert!(generate_vote_message(vote, false, &mut voting_context)
+            .unwrap()
+            .is_none());
+
+        // If we have reached wait_to_vote_slot, we should be able to vote
+        voting_context.wait_to_vote_slot = Some(1);
+        assert!(generate_vote_message(vote, false, &mut voting_context)
+            .unwrap()
+            .is_some());
+
+        // Empty authorized voter keypairs to simulate non voting node
+        voting_context.authorized_voter_keypairs = Arc::new(std::sync::RwLock::new(vec![]));
+        let vote = Vote::new_skip_vote(5);
+        // For non-voting nodes, we just return Ok(None)
+        assert!(generate_vote_message(vote, false, &mut voting_context)
+            .unwrap()
+            .is_none());
+
+        // Recover correct value to vote again
+        voting_context.authorized_voter_keypairs = Arc::new(RwLock::new(vec![Arc::new(
+            validator_keypairs[my_index].vote_keypair.insecure_clone(),
+        )]));
+        assert!(generate_vote_message(vote, false, &mut voting_context)
+            .unwrap()
+            .is_some());
+
+        // Wrong identity keypair
+        voting_context.identity_keypair = Arc::new(Keypair::new());
+        let vote = Vote::new_notarization_vote(6, Hash::new_unique());
+        assert!(generate_vote_message(vote, true, &mut voting_context)
+            .unwrap()
+            .is_none());
```

Comment:
Do these types of mismatched keys not return VoteError's? Do we have tests to cover the different types of VoteErrors?

Reply:
Our theory is that these can all happen on a hot spare when validator operators are busy configuring their hot spare validators correctly. So right now it's all INFO logs, maybe we should upgrade the logs to warning/error instead?